File: 2025/20250128/소스3.swift
//https://school.programmers.co.kr/learn/courses/30/lessons/92334?language=swift
//신고 결과 받기

import Foundation

func solution(_ id_list:[String], _ report:[String], _ k:Int) -> [Int] {
    
    // init
    var incomingEdges: [String: Set<String>] = [:]
    for log in report { 
        let arr = log.split(separator: " ").map{String($0)}
        let agent = arr[0]
        let target = arr[1]
        incomingEdges[target, default: Set<String>()].update(with: agent)
    }
    
    var mailCount: [String: Int] = [:]
    for (subject, neighbors) in incomingEdges { 
        if k <= neighbors.count {
            for neighbor in neighbors { 
                mailCount[neighbor, default: 0] += 1
            }
        }
    }
    
    var answer: [Int] = []
    for id in id_list { 
        answer.append(mailCount[id, default: 0])
    }
    
    return answer
}

File: 2025/20251029/소스3.js
//https://school.programmers.co.kr/learn/courses/30/lessons/68645
//삼각 달팽이

function solution(n) {
    const drs = [[1, 0], [0, 1], [-1, -1]] // 이렇게 반복함
    var dr_i = 0
    const matrix = []
    for (i = 0; i< n; i++ ) { 
        matrix.push(Array(i+1).fill(0))
    }
    
    const target = (1+n) * n / 2
    var count = 1
    var pos = [-1, 0]
    
    while (count <= target) {
        const next_pos = [pos[0]+drs[dr_i][0], pos[1]+drs[dr_i][1]]
        if (next_pos[0] < 0 || next_pos[0] >= n || next_pos[1] < 0 || next_pos[1] >= matrix[next_pos[0]].length ) { 
            dr_i = (dr_i + 1) % 3
            continue
        }
        if (matrix[next_pos[0]][next_pos[1]] != 0) { 
            dr_i = (dr_i + 1) % 3
            continue           
        }
        matrix[next_pos[0]][next_pos[1]] = count
        pos = next_pos
        count += 1
    }

    return matrix.flat()
}